### ISSUE TITLE
Optional async for functions zle-line-init and zle-keymap-select

### ISSUE THREAD

**Issue by Felixoid** [woefe/git-prompt.zsh#9]
Hey, Wolfgang.

After the recent update of the plugin to v2.0.0 the `$(gitprompt)` has stopped working after the command execution. It works only after vim mode switching. That's a screenshot of what do I mean:  
![screen_2019-05-19_13-16-37_window](https://user-images.githubusercontent.com/3025537/57981503-a2d2ec00-7a38-11e9-844c-9e7a18eac543.png) 

Here is my [functions](https://github.com/Felixoid/dotfiles/blob/985ac13ec270ca024ffeb9ba70e19c7cdac9a241/tag-zsh/zshrc#L30-L50)

Maybe is it still would be an option to disable async? Or do you have an idea about how to handle it properly in my case?

**upd**: I've already tried to replace `GIT_STATUS` with `gitprompt` in PS1, didn't help:

```diff
diff --git a/tag-zsh/zshrc b/tag-zsh/zshrc
index d3785db..a6342b4 100644
--- a/tag-zsh/zshrc
+++ b/tag-zsh/zshrc
@@ -25,6 +25,10 @@ if [ -r ~/.config/zsh/git-prompt.zsh ]; then
   source ~/.config/zsh/git-prompt.zsh
   _GIT_PROMPT_ENABLED=true
   ZSH_GIT_PROMPT_SHOW_STASH=true
+else
+  gitprompt() {
+    :
+  }
 fi
 
 function zle-line-init zle-keymap-select {
@@ -37,12 +41,11 @@ function zle-line-init zle-keymap-select {
     local USER_PROMPT="%F{$USER_FG}%n%f@%F{$HOST_FG}%m%f"
     [ "${SSH_CLIENT}" ] && local SSH_PROMPT="%F{yellow}[ssh from ${SSH_CLIENT%% *}]%f "
     [ "${VIRTUAL_ENV}" ] && local VENV="%F{green}venv=($(basename $VIRTUAL_ENV))%f "
-    [ "${_GIT_PROMPT_ENABLED}" ] && local GIT_STATUS="$(gitprompt) "
     case $KEYMAP in
       viins|main) local VI="%F{2}%(!.#.>)$(prompt_char)>%f" ;;
       vicmd)      local VI="%F{3}<$(prompt_char)%(!.#.<)%f" ;;
     esac
-    export PS1="┌%(?..%F{1})[%D{%H:%M:%S}]%f %(!.${ROOT_PROMPT}.${USER_PROMPT}):%~ ${VENV}${SSH_PROMPT}${GIT_STATUS}"$'\n'"└${VI} "
+    export PS1="┌%(?..%F{1})[%D{%H:%M:%S}]%f %(!.${ROOT_PROMPT}.${USER_PROMPT}):%~ ${VENV}${SSH_PROMPT}$(gitprompt)"$'\n'"└${VI} "
     RPROMPT="%(?..%F{1}[e:%?]%f)"
     zle reset-prompt
     # this line fixes the issue with errasing last line in vim mode by Alt+A or Alt+I
```

**Comment by woefe** [comment]
You can use `_zsh_git_prompt_git_status` directly, which is not asynchronous. E.g.
```
GIT_STATUS="$(_zsh_git_prompt_git_status)"
```
If this helps, I could expose this function under a nicer name and document it properly.

Is there any particular reason why you set your PS1 in zle-line-init? I set the prompt only once in my [.zshrc](https://github.com/woefe/dotfiles/blob/master/zsh/.zshrc#L276) and use zle-line-init only for setting the prompt into insert mode in my [vi-mode plugin](https://github.com/woefe/dotfiles/blob/master/zsh/.zsh-plugins/vi-mode.plugin.zsh).

If you want, you can check out my config and see how you like my vi-mode implementation. And then try to replicate it to your config. You can temporarily try my config with following commands:
```
cd /to/some/tmp/dir
git clone --depth=1 --recursive https://github.com/woefe/dotfiles.git dotfiles-woefe
cd dotfiles-woefe/zsh/
HOME=$PWD zsh -f
source .zshrc  # ignore the error that .aliases is missing
```

**Comment by Felixoid** [comment]
`_zsh_git_prompt_git_status` seems to work

I use it to draw the same PS on leave and enter normal vi-mode. Most probably I'm doing it wrong because that was the first thing which I've done after a moving to zsh.

Good to mention: if I use `"$(gitprompt)"` in PS1 it doesn't work in the way from the topic. But after replacing it with `'$(gitprompt)'` it works even in the async way.

So, on one hand, it would be a good idea to implement a blocking way. On another hand, maybe just explicit mentioning that it should be `'$(gitprompt)'` in single quotes would be enough.

Thank you for help, my problem is solved.

There is a [cast](https://asciinema.org/a/bYVMdyZi8JV0DrwUiMVcbBF6G) of my prompt behavior example together with your dotfiles.

**Comment by woefe** [comment]
> Good to mention: if I use "$(gitprompt)" in PS1 it doesn't work in the way from the topic. But after replacing it with '$(gitprompt)' it works even in the async way.

Oh yeah of course. I didn't think about that at all. The problem with double qotes is that the subsitution happens when you set PS1, which is too early. I think documenting it is the way to go. If more problems arise async, I will think about implementing a blocking option.

**Comment by woefe** [comment]
Fixed in #faae62c

**Comment by woefe** [comment]
Btw. there is now an option to disable async. https://github.com/woefe/git-prompt.zsh#disable-async
